File: stress/workloads/w1_stateless.py
from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass


@dataclass(frozen=True)
class W1AResult:
    tasks_total: int
    tasks_completed: int
    work_done: int
    duration_s: float


def _cpu_work(units: int, seed: int) -> int:
    """
    Deterministic CPU-bound work. Returns checksum so the loop isn't "empty".
    """
    h = hashlib.sha256(str(seed).encode("utf-8")).digest()
    acc = 0
    for i in range(units):
        h = hashlib.sha256(h + i.to_bytes(4, "little")).digest()
        acc ^= int.from_bytes(h[:4], "little")
    return acc


def run_w1a(tasks: int, work_units_per_task: int, seed: int) -> W1AResult:
    """
    Stateless workload: N independent tasks, deterministic work.
    """
    t0 = time.time()
    completed = 0
    checksum = 0

    for i in range(tasks):
        sub_seed = (seed * 1_000_003 + i) & 0xFFFFFFFF
        try:
            checksum ^= _cpu_work(work_units_per_task, sub_seed)
            completed += 1
        except Exception:
            # Stateless tasks: failure means "didn't complete"
            pass

    dt = time.time() - t0

    return W1AResult(
        tasks_total=tasks,
        tasks_completed=completed,
        work_done=completed,
        duration_s=dt,
    )

File: stress-ref/src/workloads/w1_stateless.rs
use sha2::{Digest, Sha256};
use std::time::Instant;

#[derive(Debug, Clone)]
pub struct W1aResult {
    pub tasks_total: u64,
    pub tasks_completed: u64,
    pub work_done: u64,
    pub duration_s: f64,
}

/// Deterministic CPU-bound work using SHA256 chain.
fn cpu_work(units: u64, seed: u64) -> u64 {
    let mut h = Sha256::digest(seed.to_le_bytes());
    let mut acc: u64 = 0;
    for i in 0..units {
        let mut hasher = Sha256::new();
        hasher.update(&h);
        hasher.update(&i.to_le_bytes());
        h = hasher.finalize();
        acc ^= u64::from_le_bytes(h[..8].try_into().unwrap());
    }
    acc
}

/// W1-A: Stateless independent task processing.
pub fn run_w1a(tasks: u64, work_units_per_task: u64, seed: u64) -> W1aResult {
    let start = Instant::now();
    let mut completed = 0u64;

    for i in 0..tasks {
        let sub_seed = seed.wrapping_mul(1_000_003).wrapping_add(i);
        let _ = cpu_work(work_units_per_task, sub_seed);
        completed += 1;
    }

    W1aResult {
        tasks_total: tasks,
        tasks_completed: completed,
        work_done: completed,
        duration_s: start.elapsed().as_secs_f64(),
    }
}
